### PR TITLE
ddl: implement backoff pool for TiFlash Cluster Management

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #32254

Problem Summary:
This PR is part of batch setting TiFlash replica in one database. It sorts out big tables that takes long time to be available into a backoff pool, in order to avoid checking every tick.

### What is changed and how it works?
When a table is to be checked for availability, we will firstly find if it is in backoff pool, and is determined not to be handled in this round.
When a table is checked to be not available for the first time, it will be added into backoff pool.
When a table is available, it will be removed from backoff pool.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
implement backoff pool for TiFlash Cluster Management
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

File: ddl/ddl_tiflash_api.go [hunk 1]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
 type TiFlashManagementContext struct {
 	TiFlashStores             map[int64]helper.StoreStat
 	HandlePdCounter           uint64
 	UpdateTiFlashStoreCounter uint64
 	UpdateMap                 map[int64]bool
+	Backoff                   *PollTiFlashBackoffContext
+}
+
+// Tick will first check increase Counter.
+func (b *PollTiFlashBackoffContext) Tick(ID int64) (bool, bool, int) {
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:123]: Please describe the three return values.

File: ddl/ddl_tiflash_api.go [hunk 2]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
+	e.TotalCounter += 1
+	return growed, true, e.TotalCounter
+}
+
+func (e *PollTiFlashBackoffElement) limit() int {
+	return int(e.Threshold)
+}
+
+// NeedGrow returns if we need to grow
+func (e *PollTiFlashBackoffElement) NeedGrow() bool {
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:139]: Could we remove this function? It only used in one place and only one line.

File: ddl/ddl_tiflash_api.go [hunk 3]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
+	return e.Counter >= e.limit()
+}
+
+func (e *PollTiFlashBackoffElement) doGrow(b *PollTiFlashBackoffContext) {
+	defer func() {
+		e.Counter = 0
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:145]: In this function only one place need to return. So I think we can put it in [line154](https://github.com/pingcap/tidb/pull/32317/files#diff-dca9a62898accc180490d462575cc0eae471c50e76c7348373075d5944addd93R154) directly.

File: ddl/ddl_tiflash_api.go [hunk 4]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
 	IsPartition    bool
 }
 
+// TiFlashTick is type for backoff threshold.
+type TiFlashTick float64
+
+// PollTiFlashBackoffElement records backoff for each TiFlash Table.
+// When `Counter` reached `Threshold`, `Threshold` shall grow.
+type PollTiFlashBackoffElement struct {
+	Counter      int
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:66]: What is this counter? Can you explain it?

File: ddl/ddl_tiflash_api.go [hunk 5]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
+
+// Get returns pointer to inner PollTiFlashBackoffElement.
+// Only exported for test.
+func (b *PollTiFlashBackoffContext) Get(ID int64) (*PollTiFlashBackoffElement, bool) {
+	res, ok := b.elements[ID]
+	return res, ok
+}
+
+// Put will record table into backoff pool, if there is enough room, or returns false.
+func (b *PollTiFlashBackoffContext) Put(ID int64) (*PollTiFlashBackoffElement, bool) {
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:181]: Why do we need to return `*PollTiFlashBackoffElement` here?

File: ddl/ddl_tiflash_api.go [hunk 6]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
+	if !ok {
+		return false, false, 0
+	}
+	growed := e.MaybeGrow(b)
+	e.Counter += 1
+	e.TotalCounter += 1
+	return growed, true, e.TotalCounter
+}
+
+func (e *PollTiFlashBackoffElement) limit() int {
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:134]: Could we remove this function? I think there are a lot of functions here.

File: ddl/ddl_tiflash_api.go (hunk 4 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_api.go:66]: I will compose a clearer comment

File: ddl/ddl_tiflash_api.go (hunk 5 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_api.go:181]: Seem to be useless currently, gonna remove.

File: ddl/ddl_tiflash_api.go (hunk 2 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_api.go:139]: I need to export this for testing.

File: ddl/ddl_tiflash_api.go (hunk 6 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_api.go:134]: intergrated into NeedGrow

File: ddl/ddl_tiflash_api.go [hunk 7]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
+func NewPollTiFlashBackoffElement() *PollTiFlashBackoffElement {
+	return &PollTiFlashBackoffElement{
+		Counter:      0,
+		Threshold:    PollTiFlashBackoffMinTick,
+		TotalCounter: 0,
+	}
+}
+
+// PollTiFlashBackoffContext is a collection of all backoff states.
+type PollTiFlashBackoffContext struct {
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:82]: The field names are a little confusing. Maybe we could refer to `tikv.Backoffer`?

File: ddl/ddl_tiflash_api.go [hunk 8]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
 	IsPartition    bool
 }
 
+// TiFlashTick is type for backoff threshold.
+type TiFlashTick float64
+
+// PollTiFlashBackoffElement records backoff for each TiFlash Table.
+// `Counter` increases every `Tick`, if it reached `Threshold`, it will be reset to 0 while `Threshold` grows.
+// `TotalCounter` records total `Tick`s this element has since created.
```
Comment (zimulala) [inline ddl/ddl_tiflash_api.go:65]: Why do I feel like these two fields are the same usage? [This](https://github.com/pingcap/tidb/pull/32317/files#diff-dca9a62898accc180490d462575cc0eae471c50e76c7348373075d5944addd93R134-R135) is the only place they do update it.

File: ddl/ddl_tiflash_api.go (hunk 2 above)
Comment (zimulala) [reply ddl/ddl_tiflash_api.go:141]: Got it.

File: ddl/ddl_tiflash_api.go [hunk 9]
```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
...
+
+// Tick will first check increase Counter.
+func (b *PollTiFlashBackoffContext) Tick(ID int64) (bool, bool, int) {
+	e, ok := b.Get(ID)
+	if !ok {
+		return false, false, 0
+	}
+	growed := e.MaybeGrow(b)
+	e.Counter += 1
+	e.TotalCounter += 1
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply ddl/ddl_tiflash_api.go:135]: Counter is necessary to backoff, however, from TotalCounter we can induce how many time this table has been pending, thus we can do much rate limit according to table size.

File: ddl/ddl_tiflash_api.go [hunk 10]
```diff
@@ -258,6 +400,11 @@ func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContex
 		// We only check unavailable tables here, so doesn't include blocked add partition case.
 		if !available {
 			allReplicaReady = false
+			grown, inqueue, _ := pollTiFlashContext.Backoff.Tick(tb.ID)
+			if inqueue && !grown {
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply ddl/ddl_tiflash_api.go:404]: Sleeps are in PollTiFlashRoutine, this is a plug-in backoffer.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**